dig_test.go

```diff
@@ -941,6 +942,1029 @@ func TestEndToEndSuccess(t *testing.T) {
 	})
 }
 
+func TestChildren(t *testing.T) {
+	t.Parallel()
+
+	t.Run("pointer constructor", func(t *testing.T) {
+		c := New()
+
```

Comment:
I don't think we need to copy all the end-to-end tests. We can parameterize
them on how they do provides and invokes instead.

So we move the existing tests into a `testEndToEnd` helper function.

```
func testEndToEnd(
    t *testing.T,
    provide func(interface{}, ...ProvideOption) error,
    invoke func(interface{}, ...InvokeOption) error,
) {
    ...
}
```

Replacing all `c.Provide` and `c.Invoke` calls in those tests with calls to
the `provide` and `invoke` arguments. (The `c := New()` can all be dropped.)

Then in the actual `TestEndToEnd` test, we can do,

```
func TestEndToEnd(t *testing.T) {
    t.Run("root", func(t *testing.T) {
        c := New()
        testEndToEnd(t, c.Provide, c.Invoke)
    })

    t.Run("provide in parent, invoke in child", func(t *testing.T) {
        parent := New()
        child := parent.Child("child")
        testEndToEnd(t, parent.Provide, child.Invoke)
    })

    t.Run("provide in child, invoke in parent", func(t *testing.T) {
        parent := New()
        child := parent.Child("child")
        testEndToEnd(t, child.Provide, parent.Invoke)
    })
}
```